p2p/README.md

```diff
@@ -0,0 +1,221 @@
+- TCP/IP transport
+- Multiplexing of different message types
+- Encrypted connections
+- DOS mitigation
+- Peer configuration and management
+- Peer discovery and churn
+- Message dissemination via gossip
+
+## Core Components
+
+### P2P
+
+The main entry point for the P2P system. It manages the overall P2P network functionality, including:
+- Starting and stopping the P2P service
+- Listening for inbound peers
+- Dialing outbound peers
+- Managing peer connections
+- Handling peer book exchanges
+- Filtering bad IPs and countries
+
+### PeerSet
+
+Manages the active set of connected peers. It provides:
+- Adding and removing peers
+- Tracking inbound and outbound connections
+- Managing peer reputation
+- Sending messages to peers
+- Handling must-connect peers
+
+### PeerBook
+
+Maintains a persistent list of potential peers. It provides:
+- Storing and retrieving peer information
+- Managing peer churn
+- Exchanging peer information with other nodes
+- Tracking failed connection attempts
+- Persisting peer data to disk
+
+### MultiConn
+
+Represents a multiplexed connection to a peer. It provides:
+- Multiple independent bi-directional communication channels
+- Rate-limited message sending and receiving
+- Ping/pong keep-alive mechanism
+- Error handling and reporting
+
+### EncryptedConn
+
+Handles the encrypted communication with peers. It implements:
+- ECDH key exchange for establishing shared secrets
+- HKDF for deriving encryption keys
+- ChaCha20-Poly1305 AEAD for message encryption and authentication
+- Handshake protocol for peer authentication
+
+### Stream
+
+Represents a single communication channel within a MultiConn. It provides:
+- Message queuing and sending
+- Message assembly from packets
+- Inbox for received messages
+
+## Sequence Diagram
+
+The following sequence diagram illustrates the core interactions in the P2P package:
+
+```mermaid
+sequenceDiagram
+    participant App as Application
+    participant P2P as P2P
+    participant PS as PeerSet
+    participant PB as PeerBook
+    participant Peer as Remote Peer
+
+    %% Connection flow
+    Note over P2P,Peer: New peer connects
+    Peer->>P2P: Connection request
+    P2P->>P2P: Security handshake
+    P2P->>PS: Add to active peers
+    
+    %% Message flow
+    App->>P2P: Send message
+    P2P->>PS: Forward to peer
+    PS->>Peer: Send encrypted message
+    
+    Peer->>P2P: Receive message
+    P2P->>App: Deliver to application
+    
+    %% Gossip protocol
+    App->>P2P: Broadcast message
+    P2P->>PS: Send to all peers except sender
+    PS->>Peer: Forward message
+    
+    %% Peer discovery
+    P2P->>P2P: Request peer list
+    P2P->>PS: Send request to peers
+    PS->>Peer: Forward request
+    
+    Peer->>P2P: Send peer list
+    P2P->>PB: Update peer book
+```
+
+## Technical Details
+
+### Multiplexing
+
+The P2P system uses multiplexing to allow multiple independent communication channels over a single TCP connection. This is achieved by:
```

Comment:
multiplexing is also an HTTP feature of http/2 which is not found in our http implementation, it would be good to clarify here that the multiplexing that is happening is not at the TCP level but at the application level to avoid confusion

Reply:
Done